pypyr/steps/contextclear.py

```diff
@@ -0,0 +1,46 @@
+"""pypyr step that clears context values.
+
+This is handy if you run the same step multiple times in a pipeline and you
+don't want the previously assigned context settings for that step to impact the
+next iteration.
+"""
+import pypyr.log.logger
+
+# logger means the log level will be set correctly
+logger = pypyr.log.logger.get_logger(__name__)
+
+
```

Comment:
You land on a pretty interesting topic :-) There's already an evolving narrative here: the 1st pypyrs literally took _anything_ as context - string, dict, list, bool, file object, take your pick. I then firmed up on making dict the a containing type so you can still put _anything_ in it, but that anything exists in the context of the context dict, as it were, rather than altering the context type itself.

First, where I agree: the validations currently are not as streamlined as they prob should be and they are indeed a bit of a grab-bag of "stuff". I'm not happy with them yet.

The current Context() class, exists pretty much solely to allow various validations against the context. While my intentions have been good, I am not yet convinced that the pattern(s) for validation especially compound steps (e.g *env* and *tar*") is as convincingly clean & minimalist as I'd like them to be - it doesn't feel like I've landed on "the one true way" yet. Currently it's sort of what I can best describe as a "reverse guard" pattern of sorts (for env and tar). No doubt these various relatively chaotic validations on the steps is what unsettled you enough to think about tightening it up - it's already biting us, really.

That said: if anything, I'm validating _too much_, and it's arguably not all that pythonic - EAFP. I do have a rationale, which is that I want the error messages to be "step blah is looking for 'blahkey' in the pypyr context but it's not specified" rather than "KeyError: 'blahkey' not found.", but rationale or not, I can and should do better.

> declarative way of controlling what ought to be on the context 

for declaring a validation on the context, hell yes, 100%! In fact, that's what something like this is trying to do (single line declare of the minimum set your context needs at a point in time):
```    
get_info, set_info, unset_info = context.keys_of_type_exist(
        ('envGet', dict),
        ('envSet', dict),
        ('envUnset', list)
    )
```

If what you mean, however, is to make it more OO style class-y and control the allowed elements more strictly in the sense of a schema, deffo ixnay to that. You'll see why when the pypyr-aws plugin lands sometime soon - dealing w complex and nested structures (say, for a REST call w json) - is actually a doozy because of this freedom. Check the json here - http://docs.aws.amazon.com/cli/latest/reference/ecs/register-task-definition.html - you can actually literally pretty much bang that into the context as it is, nested elements and all, and use it like any python dict-containing-dicts in the code.

This power wasn't an accident - I've been after since the beginning as a First Principle to make something where I don't need to translate between a "pypyr format" and a "aws format" - it's bad enough having to figure out the actual api end-point you want to use, adding some sort of "canonical" schema in the middle pypyr-side to represent that api call, when what you're actually interested in is the api call itself is actually exactly what I wanted to avoid from day 1.

(as an aside and nothing to do with your original comment, the use of the word "canonical" in enterprise software is actually the height of arrogance - it basically says our little system and its little demands are more important than anything else our little system might be talking to)

> and extraneous context is not there.

deffo not. . . extraneous content doesn't harm anyone except the CPU and the RAM and billings in the cloud 😆 .

In closing, philosophically, the context and pypyr embraces dynamic coding styles - I'm viewing schemafying it as almost what typescript is trying to do to javascript. There are trade-offs (validation is harder to get right, run time errs because of mystery things in context), but for that you get huge awesome power (extensibility, adaptability).